mlflow/entities/run_data.py

```diff
@@ -2,66 +2,71 @@
 from mlflow.entities.metric import Metric
 from mlflow.entities.param import Param
 from mlflow.entities.run_tag import RunTag
-from mlflow.protos.service_pb2 import RunData as ProtoRunData
+from mlflow.protos.service_pb2 import RunData as ProtoRunData, Param as ProtoParam,\
+    RunTag as ProtoRunTag
 
 
 class RunData(_MLflowObject):
     """
     Run data (metrics and parameters).
     """
     def __init__(self, metrics=None, params=None, tags=None):
-        self._metrics = []
-        self._params = []
-        self._tags = []
-        if metrics is not None:
-            for m in metrics:
-                self._add_metric(m)
-        if params is not None:
-            for p in params:
-                self._add_param(p)
-        if tags is not None:
-            for t in tags:
-                self._add_tag(t)
+        """
+        Construct a new :py:class:`mlflow.entities.RunData` instance.
+        :param metrics: Dictionary of metric key (string) to :py:class:`mlflow.entities.Metric`.
+        :param params: Dictionary of param key (string) to param value (string).
+        :param tags: Dictionary of tag key (string) to tag value (string).
+        """
+        self._metrics = metrics or {}
+        self._params = params or {}
+        self._tags = tags or {}
 
     @property
     def metrics(self):
-        """List of :py:class:`mlflow.entities.Metric` for the current run."""
+        """
+        Dictionary of string key -> :py:class:`mlflow.entities.Metric` for the current run.
+        For each metric key, the maximum metric value at the maximum timestamp is returned.
+        """
         return self._metrics
 
     @property
     def params(self):
-        """List of :py:class:`mlflow.entities.Param` for the current run."""
+        """Dictionary of param key (string) -> param value (string) for the current run."""
         return self._params
 
     @property
     def tags(self):
-        """List of :py:class:`mlflow.entities.RunTag` for the current run."""
+        """Dictionary of tag key (string) -> tag value (string) for the current run."""
         return self._tags
 
     def _add_metric(self, metric):
         if isinstance(metric, dict):
             metric = Metric(metric['key'], metric['value'], metric['timestamp'])
-        self._metrics.append(metric)
+        self._metrics[metric.key] = metric
 
     def _add_param(self, param):
         if isinstance(param, dict):
             param = Param(param['key'], param['value'])
-        self._params.append(param)
+        self._params[param.key] = param.value
```

Comment:
Good question, it's because metrics have additional fields besides their value (like timestamp, and in the near future, x coordinates)

Reply:
I guess I was more thinking if it would make sense to keep the other ones as map to Param / Tag instead of directly pointing to the value for the sake of consistency.

Reply:
Yeah I think that's a fair point, this is actually a design point that probably should have been discussed more broadly before starting implementation. I guess the questions are:

1) Are there significant ease-of-use gains from being able to access tag/param values within a flat dictionary (e.g. `param_val = run.data.params["alpha"]`) vs having to perform an extra access to get to the value (`param_val = run.data.params["alpha"].value`)? If so, are these ease-of-use gains worth a) the inconsistency of having metrics be a dict of string -> Metric object, and b) the inflexibility of not being able to add new fields to params/tags without breaking API changes?